### ISSUE TITLE
【Feature Request】前向重计算函数在 use_reentrant == True时支持以关键字参数的方式传入 Tensor

### ISSUE THREAD

**Issue by MarioLulab** [PaddlePaddle/Paddle#62363]
### 一、任务目标
> 请描述你正在做的项目是什么，如模型、论文、项目是什么？

前向重计算是将在前向计算时，除了小部分必须存储在内存中的张量外，其他中间结果都将被删除；在反向计算中，首先重新计算一遍前向算子，以获得中间结果，再运行反向算子。简而言之，前向重计算出网络中间的激活值，从而达到节省显存的目的。用户使用 Paddle 定义网络时，可以使用 `recompute` 函数定义网络中进行重计算的层。`recompute` 函数内部（`recompute` 函数位于 `python/paddle/distributed/fleet/recompute/recompute.py`）由 `use_reentrant` 参数控制，会采用两种方法实现重计算功能：`use_reentrant == True` 时会使用 `PyLayer` 来实现。但 `PyLayer` 目前不支持以关键字参数的形式传入 Tensor 类型参数（因为以 dict 形式传入的 Tensor 无法正确的创建反向节点、反向边） 。
为了提升分布式训练的易用性，你的任务是：支持 `recompute` 函数在 `use_reentrant == True` 时能以关键字参数的方式传入 tensor，同时要求启用重计算后性能不下降。

### 二、需求场景
> 请描述你的项目中为什么需要用此功能

前向重计算是将在前向计算时，除了小部分必须存储在内存中的张量外，其他中间结果都将被删除；在反向计算中，首先重新计算一遍前向算子，以获得中间结果，再运行反向算子。简而言之，前向重计算出网络中间的激活值，从而达到节省显存的目的。用户使用 Paddle 定义网络时，可以使用 `recompute` 函数定义网络中进行重计算的层。以下是 [Paddle 官方给的样例](https://www.paddlepaddle.org.cn/documentation/docs/zh/develop/guides/06_distributed_training/data_parallel/recompute_cn.html)

```python
import paddle
from paddle.distributed.fleet.utils import recompute
import random
paddle.seed(2023)
def get_fc_block(block_idx, input_size, is_last=False):
    block_name = "block_" + str(block_idx)
    block = paddle.nn.Sequential(
        (block_name + "_fc_0", paddle.nn.Linear(input_size, input_size, bias_attr=False)),
        (block_name + "_dropout", paddle.nn.Dropout(p=0.5)),
        (block_name + "_relu_1", paddle.nn.ReLU()),
        (block_name + "_fc_1", paddle.nn.Linear(input_size, input_size, bias_attr=False)),
        (block_name + "_relu_2", paddle.nn.ReLU()),
    )
    if is_last:
        block.add_sublayer(
            block_name + "_fc_2",
            paddle.nn.Linear(
                input_size, 1, bias_attr=False
            )
        )
    else:
        block.add_sublayer(
            block_name + "_fc_2",
            paddle.nn.Linear(input_size, input_size, bias_attr=False)
        )
    return block

class Naive_fc_net(paddle.nn.Layer):
    def __init__(self, input_size=10,
                recompute_blocks=[1, 3],
                recompute_kwargs={}):
        super().__init__()
        self.recompute_blocks = recompute_blocks
        self.recompute_kwargs = recompute_kwargs
        self.runfunc0 = get_fc_block(0, input_size, is_last=False)
        self.runfunc1 = get_fc_block(1, input_size, is_last=False)
        self.runfunc2 = get_fc_block(2, input_size, is_last=False)
        self.runfunc3 = get_fc_block(3, input_size, is_last=False)
        self.runfunc4 = get_fc_block(4, input_size, is_last=True)
        self.total_func = [self.runfunc0, self.runfunc1, self.runfunc2, self.runfunc3, self.runfunc4]
    def forward(self, inputs):
        nums = len(self.total_func)
        for i in range(nums):
            if i in self.recompute_blocks:
                inputs = recompute(self.total_func[i], inputs, **{"preserve_rng_state": True})       #  <---- 指定重计算网络层
            else:
                inputs = self.total_func[i](inputs)
        return inputs

def run_model(cuda_state, recompute_block=[], recompute_kwargs={}):
    gen = paddle.seed(10)
    gen.manual_seed(10)
    random.seed(10)
    if cuda_state:
        paddle.set_cuda_rng_state(cuda_state)
    batch_size, input_size = 1, 10
    model = Naive_fc_net(
        input_size,
        recompute_blocks=recompute_block,
        recompute_kwargs=recompute_kwargs)
    optimizer = paddle.optimizer.SGD(learning_rate=0.01, parameters=model.parameters())
    loss_ = []
    param_ = []
    grad_ = []
    for _ in range(5):
        x = paddle.rand(shape=[batch_size, input_size], dtype="float32")
        y_pred = model(x)
        loss = y_pred.mean()
        loss_.append(loss.item())
        loss.backward()
        optimizer.step()
        param_.append(model.parameters()[9])
        grad_.append(model.parameters()[3]._grad_ivar())
        optimizer.clear_grad()
    return loss_, param_, grad_

cuda_state = paddle.get_cuda_rng_state()
# without recompute
loss_ref, param_ref, grad_ref = run_model(
    cuda_state, recompute_block=[]
)

loss, param, grad = run_model(cuda_state, recompute_block=[1, 2])
print("normal_loss: {}, recompute_loss: {}".format(loss_ref, loss))
# The result of the recompute_loss should be the same as the normal_loss.
normal_loss: [0.0018744759727269411, 0.0, 0.035971127450466156, 0.0, 0.0], recompute_loss: [0.0018744759727269411, 0.0, 0.035971127450466156, 0.0, 0.0]
```

在默认情况下 `recompute` 函数的 `use_reentrant` 属性为 `True`。目前，`recompute` 函数在 `use_reentrant == True`
 时不支持以 dict 形式传入 Tensor，因为 `PyLayer` 目前不支持以关键字参数的形式传入 Tensor 类型参数（因为以关键词参数形式传入的 Tensor 无法正确的创建反向节点、反向边）

### 三、功能描述

支持 `recompute` 函数在 `use_reentrant == True` 时能以关键字参数的方式传入 tensor，同时要求启用重计算后性能不下降。以下是一个功能支持后组网的样例：
```python
class Naive_fc_net(paddle.nn.Layer):
    def __init__(self, input_size=10,
                recompute_blocks=[1, 3],
                recompute_kwargs={}):
        super().__init__()
        self.recompute_blocks = recompute_blocks
        self.recompute_kwargs = recompute_kwargs
        self.runfunc0 = get_fc_block(0, input_size, is_last=False)
        self.runfunc1 = get_fc_block(1, input_size, is_last=False)
        self.runfunc2 = get_fc_block(2, input_size, is_last=False)
        self.runfunc3 = get_fc_block(3, input_size, is_last=False)
        self.runfunc4 = get_fc_block(4, input_size, is_last=True)
        self.total_func = [self.runfunc0, self.runfunc1, self.runfunc2, self.runfunc3, self.runfunc4]
    def forward(self, inputs):
        nums = len(self.total_func)
        for i in range(nums):
            if i in self.recompute_blocks:
                inputs = recompute(self.total_func[i], **{"preserve_rng_state": True, 'x' : inputs})
            else:
                inputs = self.total_func[i](inputs)
        return inputs
```

**Comment by MarioLulab** [comment]
**NOTE:** 当前 `recompute` 函数在 `use_reentrant == True` 时不支持以 dict 形式传入 Tensor 类型参数的根本原因是：以 dict 形式传入给 `PyLayer.apply` 方法的 Tensor 不会创建反向节点、反向边。以下提供几个关于 `PyLayer` 相关的使用事项说明：

### 一、PyLayer 实现原理
**首先是 PyLayer 的机制流程：** 

![image](https://github.com/PaddlePaddle/Paddle/assets/61354321/0faea50f-c616-4a04-b415-0757388cba76)


PyLayer 继承自core.eager.PyLayer，提供了两个需要重写的接口：

```python
class EagerPyLayer(
        with_mateclass(EagerPyLayerMeta, core.eager.PyLayer,
                       EagerPyLayerContext)):

    @staticmethod
    def forward(ctx, *args, **kwargs):
         raise NotImplementedError(
            "You must implement the forward function for PyLayer.")

    @staticmethod
    def backward(ctx, *args):
         raise NotImplementedError(
            "You must implement the backward function for PyLayer.")
```

C++ 端定义了一个 `PyLayerObject` 的结构体：

```c++
typedef struct {
  PyObject_HEAD PyObject* container;
  bool container_be_packed;
  std::shared_ptr<egr::UnPackHookBaseunpack_hook;
  PyObject* non_differentiable;
  PyObject* not_inplace_tensors;
  bool materialize_grads;
  std::vector<boolforward_input_tensor_is_duplicable;
  std::vector<boolforward_output_tensor_is_duplicable;
  std::weak_ptr<egr::GradNodePyLayergrad_node;
} PyLayerObject;
```

借助 `BindEagerPyLayer()` 函数实现了关键属性、方法的 bind：

```c++
void BindEagerPyLayer(PyObject* module) {
  auto heap_type = reinterpret_cast<PyHeapTypeObject*>(
      PyType_Type.tp_alloc(&PyType_Type, 0));
  heap_type->ht_name = ToPyObject("PyLayer");
  heap_type->ht_qualname = ToPyObject("PyLayer");
  auto type = &heap_type->ht_type;
  type->tp_name = "PyLayer";
  type->tp_basicsize = sizeof(PyLayerObject);
  type->tp_dealloc = (destructor)PyLayerDealloc;
  type->tp_methods = pylayer_methods;    // <----- 核心方法
  type->tp_getset = pylayer_properties;  // <----- 核心属性
  type->tp_new = (newfunc)PyLayerNew;
  // 省略
}
```

后端绑定的只有两个 method，一个是name()，另一个是 `apply(self, *args, **kwargs)` ，后者是 PyLayer 核心逻辑：

* 首先会解析用户传入的args和kwargs参数
* 然后调用用户的 forward函数：`outputs = PyObject_Call(forward_fn, forward_args, kwargs)` ;
* 创建反向grad_node节点：`GradNodePyLayer`
* 当用户调用`loss.backward()`时，最终会执行 `GradNodePyLayer::operator()`
  
  * 在C++端调用用户的backward函数：`auto outputs = PyObject_CallObject(backward_fn, backward_args)`;

对于反向 GradNodePyLayer，其继承自GradNodeBase，核心成员如下：

```c++
class GradNodePyLayer : public GradNodeBase {
 public:
  GradNodePyLayer(PyObject* ctx,
                  size_t bwd_in_slot_num,
                  size_t bwd_out_slot_num)
      : GradNodeBase(bwd_in_slot_num, bwd_out_slot_num) {
    ctx_ = ctx;
    Py_INCREF(ctx_);
  }
  
  private:
  PyObject* ctx_{nullptr};    //<----- 记录了Pythond端的PyLayer对象，用于获取backward函数指针
  std::vector<std::vector<phi::DenseTensorMeta>forward_outputs_meta_;
  std::vector<std::vector<paddle::platform::Place>forward_outputs_place_;
};
```

**Comment by MarioLulab** [comment]
### 二、PyLayer交互参数的简单说明
PyLayer是C++实现的Paddle机制，用于支持自定义反向。在用户使用中，发现用户对PyLayer过于“信任”，会给PyLayer传递各种格式的参数，有时候会导致反向图组网错误。以下给出简单说明：
1. apply()的输入只支持：Tensor, List of Tensor, Tuple of Tensor，其余形式的Python对象将不视为Tensor只做透传给forward()。比如List of List of Tensor就不能当成Tensor处理。只有被识别为Tensor的输入stop_gradient==False，才会创建反向节点、建立反向边
2. forward()的输出，也只支持：Tensor, List of Tensor, Tuple of Tensor，其余形式的Python对象将不视为Tensor只做透传返回给apply的调用方。只有被识别为Tensor的输出才会建立反向边，才能有梯度。
3. backward()的输入，由PyLayer机制生成，只会生成Tensor, Tuple of Tensor两种，如果前向的返回值是List of Tensor，它的梯度作为backward的输入会是Tuple of Tensor。举例，如果forward的返回值是：Tensor，int，List of Tensor, List of List of Tensor，则backward的输入会是Tensor, Tuple of Tensor
4. backward()的输出，应该与forward输入的Tensor“格式”严格一一对应
  a. 如果某个Tensor没有梯度，可以传None
  b. 如果某个List of Tensor中部分Tensor没有梯度，需要传[Tensor, None, Tensor, ...]
  c. 如果forward输入的List of Tensor的第1个Tensorstop_gradient==True，PyLayer会认为整个List都是不需要梯度的，它期望backward的返回值是None，否则报错
